### ISSUE TITLE
ASSERT_PARAM(-218959118 0), in arch_main.c at line 343 (IDFGH-6126)

### ISSUE THREAD

**Issue by ChrisEAlfred** [espressif/esp-idf#7805]
Opening - same as https://github.com/espressif/esp-idf/issues/6517. Still occurs in 4.3.1

**Comment by alexmantaut** [comment]
Same here, it happens to me on 4.2.2.

It is less frequent than it used to be, but it still happens

```
ASSERT_PARAM(-218959118 0), in arch_main.c at line 343
Guru Meditation Error: Core  0 panic'ed (Interrupt wdt timeout on CPU0). 

Core  0 register dump:
PC      : 0x40088b64  PS      : 0x00060434  A0      : 0x801395d0  A1      : 0x3ffbf8f0  
A2      : 0x00000001  A3      : 0x00000000  A4      : 0x60008048  A5      : 0x00000000  
A6      : 0x00000004  A7      : 0x3ffbdcb8  A8      : 0x80088b64  A9      : 0x3ffbf8d0  
A10     : 0x0000f2f2  A11     : 0x00000037  A12     : 0x00000014  A13     : 0xffffffff  
A14     : 0x00000000  A15     : 0xfffffffc  SAR     : 0x00000004  EXCCAUSE: 0x00000005  
EXCVADDR: 0x00000000  LBEG    : 0x40088a9c  LEND    : 0x40088aa3  LCOUNT  : 0x00000000  
Core  0 was running in ISR context:
EPC1    : 0x401ccdf3  EPC2    : 0x00000000  EPC3    : 0x00000000  EPC4    : 0x40088b64

Backtrace:0x40088b61:0x3ffbf8f0 0x401395cd:0x3ffbf910 0x40019fb5:0x3ffbf930 0x40046683:0x3ffbf960 0x40047515:0x3ffbf980 0x4008bbb1:0x3ffbf9a0 0x4008b805:0x3ffbf9c0 0x4008df12:0x3ffbf9e0 0x4008f18f:0x3ffbfa00 0x4008645d:0x3ffbfa20 0x401399f5:0x3ffb5f10 0x40139fd5:0x3ffb5f30 0x4009487e:0x3ffb5f60
Core  1 register dump:
PC      : 0x4009296d  PS      : 0x00060d34  A0      : 0x80092ad5  A1      : 0x3ffbcc00  
A2      : 0x3ffc76a8  A3      : 0x00060d20  A4      : 0x800949c4  A5      : 0x3ffbcb20  
A6      : 0x00000000  A7      : 0x3ffbc500  A8      : 0x00000015  A9      : 0x00000001  
A10     : 0x000001a4  A11     : 0x00000000  A12     : 0x00060d20  A13     : 0x00060d23  
A14     : 0x00000001  A15     : 0x3ffbc500  SAR     : 0x0000001d  EXCCAUSE: 0x00000005  
EXCVADDR: 0x00000000  LBEG    : 0x00000000  LEND    : 0x00000000  LCOUNT  : 0x00000000  

Backtrace:0x4009296a:0x3ffbcc00 0x40092ad2:0x3ffbcc20 0x40092cf0:0x3ffbcc40 0x4009487e:0x3ffbcc60
```

**Comment by AxelLin** [comment]
@Alvin1Zhang 
This issue has been reported quite long time ago, and at least 3 people reported that the fix does not really fix the issue.
https://github.com/espressif/esp-idf/issues/6517#issuecomment-891186177

**Comment by jmaha** [comment]
This ongoing issue is the reason we removed the ESP32 from a product we're developing.

**Comment by TianaESP** [comment]
Thanks for reporting the issue. 
The crash was because after the controller ran out of memory, malloc failure happens and assert is triggered. The issue is usually caused by the slow processing of the application layer, which caused the controller memory could not be released in time. We did some optimization before, however, if the latest IDF version still has the issue, please refer to the following suggestions and give it a try. 
1. Reduce time-consuming operations at the application layer, such as a large number of log printing.
2. Increase the CPU speed, for example, increase the CPU frequency to 240Mhz. If dual-core is supported, run bluedroid host and Controller on different cores.
3. If you are scanning at the same time:
Try to reduce controller memory consumption, for example, by making the following changes in the configuration:
(1) Change **BLE adv report flow control number** to 20 and **BLE adv lost event threshold value** to 100
(2) Disable **BLE full scan feature supported**
(3) Enable **BLE queue congestion check** 
(4) Disable **Report adv data and scan response individually when BLE active scan**
(5) Enable **duplicate Scan** when setting **scan parameters** to filter out duplicate broadcast packets

Meanwhile, we tried but unfortunately did not reproduce the issue yesterday. Could you please provide a tiny demo with the steps of reappearance? We will keep looking into this. Thanks.

Tiana, Espressif

**Comment by Platin21** [comment]
@TianaESP we have a similar error but 1) is not possible as we need the tasks todo the stuff they need todo. 2) Doesn't work as we have a very narrow power margin. 3) We can certainly try that but we are not using bluedroid we are using the nimble stack which seems to also use libbtdm in which this assertion seems to happen. 

May i ask what is the particular reason why there is not source for this part? As that would make debugging easier. 
(Not much of a fan of re'ing such stuff or reading xtensa assembly.. )

We can reporduce the issue with a run that takes around +3 hours. 
Not sure if there can be a minimal repro made as you guy's would still require the hardware on which we do the tests. 
We can supply stacktraces and registers if that may help. Or if you wish we can certainly setup a remote session or something in order to get this working again. This is basically a show stopper for our entire project so if this can't be fixed we have to move to a diffrent RTOS and do a lot of the work again.

**Comment by Platin21** [comment]
I'm not to sure if the controller memory is the problem as we have 4096 allocated to it so that is plenty or do you mean there is a leak inside the controller task? That would be very unfortnunate.

**Comment by AxelLin** [comment]
> Meanwhile, we tried but unfortunately did not reproduce the issue yesterday. Could you please provide a tiny demo with the steps of reappearance? We will keep looking into this. Thanks.
> 

@TianaESP 

Just to clarify about your comments above.

So you should be able to reproduce it by
- Reduce CPU speed
- Disable dual-core
- Add more logging

In additional, this issue does not happen immediately.
The device can run for hours then it suddenly hit assert.
So it's really difficult for application developers to know what could be wrong.

**Comment by Platin21** [comment]
@AxelLin in our case we have dual-core running, we have the highest frequnecy set, we have it still happening.

**Comment by TianaESP** [comment]
@Platin21  Can you please provide the specific commit id of IDF and describe what the host task did?

**Comment by Platin21** [comment]
@TianaESP 
We are using version 4.3.1 and 4.4 but it was the same on ealier builds. We didn't want to go full master as we are not sure about the stability of the other features that we use with that. Are there any recent changes that should have fixed this?

**Comment by Platin21** [comment]
I can certainly provide you a trace not sure if that will help alot though. We don't have JTAG on the board so debugging is somewhat difficult.  

````bash
[0;31mE (3332242) task_wdt: Task watchdog got triggered. The following tasks did not reset the watchdog in time:[0m
[0;31mE (3332242) task_wdt:  - IDLE (CPU 1)[0m
[0;31mE (3332242) task_wdt: Tasks currently running:[0m
[0;31mE (3332242) task_wdt: CPU 0: IDLE[0m
[0;31mE (3332242) task_wdt: CPU 1: btController[0m
[0;31mE (3332242) task_wdt: Print CPU 0 (current core) backtrace[0m


Backtrace:0x40112F9E:0x3FFBE9A00x40083299:0x3FFBE9D0 0x4018BF4F:0x3FFBC7E0 0x400D4D3D:0x3FFBC800 0x40093907:0x3FFBC820 

[0;31mE (3332242) task_wdt: Print CPU 1 backtrace[0m


Backtrace:0x40085055:0x3FFBEFB00x40083299:0x3FFBEFD0 0x40174091:0x3FFE7C20 0x40174659:0x3FFE7C40 

[0;32mI (3332399) UDS: Read data id fe20[0m
[0;32mI (3332406) UDS: Read data id fe1f[0m
[0;32mI (3332425) UDP SERVER: Received 9 bytes from 192.168.0.68:[0m
[0;32mI (3332426) UDP SERVER: JoDaHomas[0m
[0;32mI (3332430) UDP SERVER: Waiting for data[0m
[0;31mE (3337242) task_wdt: Task watchdog got triggered. The following tasks did not reset the watchdog in time:[0m
[0;31mE (3337242) task_wdt:  - IDLE (CPU 1)[0m
[0;31mE (3337242) task_wdt: Tasks currently running:[0m
[0;31mE (3337242) task_wdt: CPU 0: IDLE[0m
[0;31mE (3337242) task_wdt: CPU 1: btController[0m
[0;31mE (3337242) task_wdt: Print CPU 0 (current core) backtrace[0m


Backtrace:0x40112F9E:0x3FFBE9A00x40083299:0x3FFBE9D0 0x4018BF4F:0x3FFBC7E0 0x400D4D3D:0x3FFBC800 0x40093907:0x3FFBC820 

[0;31mE (3337242) task_wdt: Print CPU 1 backtrace[0m


Backtrace:0x40085055:0x3FFBEFB00x40083299:0x3FFBEFD0 0x40174091:0x3FFE7C20 0x40174659:0x3FFE7C40 

[0;32mI (3337321) UDP SERVER: Received 9 bytes from 192.168.0.68:[0m
[0;32mI (3337321) UDP SERVER: JoDaHomas[0m
[0;32mI (3337325) UDP SERVER: Waiting for data[0m
[0;32mI (3337647) UDS: Read data id fe20[0m
[0;32mI (3337653) UDS: Read data id fe1f[0m
[0;31mE (3342242) task_wdt: Task watchdog got triggered. The following tasks did not reset the watchdog in time:[0m
[0;31mE (3342242) task_wdt:  - IDLE (CPU 1)[0m
[0;31mE (3342242) task_wdt: Tasks currently running:[0m
[0;31mE (3342242) task_wdt: CPU 0: IDLE[0m
[0;31mE (3342242) task_wdt: CPU 1: btController[0m
[0;31mE (3342242) task_wdt: Print CPU 0 (current core) backtrace[0m


Backtrace:0x40112F9E:0x3FFBE9A00x40083299:0x3FFBE9D0 0x4018BF4F:0x3FFBC7E0 0x400D4D3D:0x3FFBC800 0x40093907:0x3FFBC820 

[0;31mE (3342242) task_wdt: Print CPU 1 backtrace[0m


Backtrace:0x40085055:0x3FFBEFB00x40083299:0x3FFBEFD0 0x40174091:0x3FFE7C20 0x40174659:0x3FFE7C40 

[0;32mI (3342356) UDP SERVER: Received 9 bytes from 192.168.0.68:[0m
[0;32mI (3342357) UDP SERVER: JoDaHomas[0m
[0;32mI (3342360) UDP SERVER: Waiting for data[0m
[0;32mI (3342928) UDS: Read data id fe20[0m
[0;32mI (3342934) UDS: Read data id fe1f[0m
[0;31mE (3347242) task_wdt: Task watchdog got triggered. The following tasks did not reset the watchdog in time:[0m
[0;31mE (3347242) task_wdt:  - IDLE (CPU 1)[0m
[0;31mE (3347242) task_wdt: Tasks currently running:[0m
[0;31mE (3347242) task_wdt: CPU 0: IDLE[0m
[0;31mE (3347242) task_wdt: CPU 1: btController[0m
[0;31mE (3347242) task_wdt: Print CPU 0 (current core) backtrace[0m


Backtrace:0x40112F9E:0x3FFBE9A00x40083299:0x3FFBE9D0 0x4018BF4F:0x3FFBC7E0 0x400D4D3D:0x3FFBC800 0x40093907:0x3FFBC820 

[0;31mE (3347242) task_wdt: Print CPU 1 backtrace[0m


Backtrace:0x40085055:0x3FFBEFB00x40083299:0x3FFBEFD0 0x40174091:0x3FFE7C20 0x40174659:0x3FFE7C40 

[0;32mI (3347382) UDP SERVER: Received 9 bytes from 192.168.0.68:[0m
[0;32mI (3347383) UDP SERVER: JoDaHomas[0m
[0;32mI (3347388) UDP SERVER: Waiting for data[0m
[0;32mI (3348169) UDS: Read data id fe20[0m
[0;32mI (3348178) UDS: Read data id fe1f[0m
ASSERT_PARAM(-218959118 0), in arch_main.c at line 343                                               <==== The reported ASSERT
Guru Meditation Error: Core  0 panic'ed (Interrupt wdt timeout on CPU0). 

Core  0 register dump:
PC      : 0x4018bf52  PS      : 0x00060334  A0      : 0x800d4d40  A1      : 0x3ffbc7e0  
A2      : 0x00000000  A3      : 0x00000001  A4      : 0x00000014  A5      : 0x000000a5  
A6      : 0x000000a5  A7      : 0x00060023  A8      : 0x801132ae  A9      : 0x3ffbc7c0  
A10     : 0x00000000  A11     : 0x00000001  A12     : 0x00000014  A13     : 0x000000a5  
A14     : 0x000000a5  A15     : 0x00060023  SAR     : 0x0000001d  EXCCAUSE: 0x00000005  
EXCVADDR: 0x00000000  LBEG    : 0x00000000  LEND    : 0x00000000  LCOUNT  : 0x00000000  


Backtrace:0x4018bf4f:0x3ffbc7e00x400d4d3d:0x3ffbc800 0x40093907:0x3ffbc820 


Core  1 register dump:
PC      : 0x40089689  PS      : 0x00060f34  A0      : 0x80173c7c  A1      : 0x3ffbeec0  
A2      : 0x00000001  A3      : 0x00000000  A4      : 0x60008048  A5      : 0x00000000  
A6      : 0x00000004  A7      : 0x3ffc21ec  A8      : 0x80089684  A9      : 0x3ffbeea0  
A10     : 0x00000000  A11     : 0x00000037  A12     : 0x00000014  A13     : 0xffffffff  
A14     : 0x00000000  A15     : 0xfffffffc  SAR     : 0x00000004  EXCCAUSE: 0x00000005  
EXCVADDR: 0x00000000  LBEG    : 0x400895bc  LEND    : 0x400895c3  LCOUNT  : 0x00000000  
````

**Comment by Platin21** [comment]
And here is the decoded addresses sadly still some ??'s:
```c
CONSOLE>xtensa-esp32-elf-addr2line.exe -e cu-offroad.elf -p -a 0x40112F9E:0x3FFBE9A00x40083299:0x3FFBE9D0 0x4018BF4F:0x3FFBC7E0 0x400D4D3D:0x3FFBC800 0x40093907:0x3FFBC820
0x40112f9e: #idf path#/esp_system/task_wdt.c:191 (discriminator 3)
0x4018bf4f: #idf path#/hal/esp32/include/hal/cpu_ll.h:183
0x400d4d3d: #idf path#/esp_system/freertos_hooks.c:63
0x40093907: #idf path#/freertos/tasks.c:3823

CONSOLE>xtensa-esp32-elf-addr2line.exe -e cu-offroad.elf -p -a 0x40085055:0x3FFBEFB00x40083299:0x3FFBEFD0 0x40174091:0x3FFE7C20 0x40174659:0x3FFE7C40
0x40085055: #idf path#/esp_system/crosscore_int.c:92
0x40174091: ??:?
0x40174659: ??:?

CONSOLE>xtensa-esp32-elf-addr2line.exe -e cu-offroad.elf -p -a 0x40112F9E:0x3FFBE9A00x40083299:0x3FFBE9D0 0x4018BF4F:0x3FFBC7E0 0x400D4D3D:0x3FFBC800 0x40093907:0x3FFBC820
0x40112f9e: #idf path#/esp_system/task_wdt.c:191 (discriminator 3)
0x4018bf4f: #idf path#/hal/esp32/include/hal/cpu_ll.h:183
0x400d4d3d: #idf path#/esp_system/freertos_hooks.c:63
0x40093907: #idf path#/freertos/tasks.c:3823

CONSOLE>xtensa-esp32-elf-addr2line.exe -e cu-offroad.elf -p -a 0x40112F9E:0x3FFBE9A00x40083299:0x3FFBE9D0 0x4018BF4F:0x3FFBC7E0 0x400D4D3D:0x3FFBC800 0x40093907:0x3FFBC820
0x40112f9e: #idf path#/esp_system/task_wdt.c:191 (discriminator 3)
0x4018bf4f: #idf path#/hal/esp32/include/hal/cpu_ll.h:183
0x400d4d3d: #idf path#/esp_system/freertos_hooks.c:63
0x40093907: #idf path#/freertos/tasks.c:3823

CONSOLE>xtensa-esp32-elf-addr2line.exe -e cu-offroad.elf -p -a 0x40085055:0x3FFBEFB00x40083299:0x3FFBEFD0 0x40174091:0x3FFE7C20 0x40174659:0x3FFE7C40
0x40085055: #idf path#/esp_system/crosscore_int.c:92
0x40174091: ??:?
0x40174659: ??:?

CONSOLE>xtensa-esp32-elf-addr2line.exe -e cu-offroad.elf -p -a 0x40085055:0x3FFBEFB00x40083299:0x3FFBEFD0 0x40174091:0x3FFE7C20 0x40174659:0x3FFE7C40
0x40085055: #idf path#/esp_system/crosscore_int.c:92
0x40174091: ??:?
0x40174659: ??:?

CONSOLE>xtensa-esp32-elf-addr2line.exe -e cu-offroad.elf -p -a 0x4018bf4f:0x3ffbc7e00x400d4d3d:0x3ffbc800 0x40093907:0x3ffbc820
0x4018bf4f: #idf path#/hal/esp32/include/hal/cpu_ll.h:183
0x40093907: #idf path#/freertos/tasks.c:3823

CONSOLE>xtensa-esp32-elf-addr2line.exe -e cu-offroad.elf -p -a 0x40089686:0x3ffbeec00x40173c79:0x3ffbeee0 0x40019fb5:0x3ffbef00 0x400466c3:0x3ffbef30 0x4008c772:0x3ffbef50 0x4008c45d:0x3ffbef70 0x4008eb6a:0x3ffbef90 0x4008fde7:0x3ffbefb0 0x40083299:0x3ffbefd0 0x40174091:0x3ffe7c20 0x40174659:0x3ffe7c40
0x40089686: ??:?
0x40019fb5: ??:0
0x400466c3: ??:0
0x4008c772: ??:?
0x4008c45d: ??:?
0x4008eb6a: ??:?
0x4008fde7: intc.c:?
0x40083299: #idf path#/freertos/port/xtensa/xtensa_vectors.S:1105
0x40174091: ??:?
0x40174659: ??:?
```

**Comment by Platin21** [comment]
In this case it was running for around 3 hours which is for our use case not that uncommon. And as i said we are willing to take some time to debug the issue but for us it's not a clear case. And we can only see that libbtdm as inside the controller task we have a failure.

**Comment by Platin21** [comment]
After this the device has a Guru Mediation Error and Restarts which kills the BLE Connection and makes it worthless for the customer.

**Comment by skyESP** [comment]
1.Coud you give the commit id of the idf you are using?
2 Could you describe the demo you are using?
We have a demo which the device do scan , adv and connect to four other devices. But we run about one day and could not reproduce the issue.

**Comment by Platin21** [comment]
@skyESP 
1:
rel4.4 https://github.com/espressif/esp-idf/tree/release/v4.4
rel4.3 https://github.com/espressif/esp-idf/tree/release/v4.3
This report was from a run with version 4.4 

2:
Just a normal BLE process with RPA Enabled.
As in start with adv then do a connect then we have GATT Attributes that we retrieve every 5 seconds. Our process also does do a scan for sensors but that is not active during the tests.

**Comment by skyESP** [comment]
During the test, Does the Adv is still continue? If it does,what's the frequence?

**Comment by Platin21** [comment]
@skyESP 
1. No it doesn’t continue 
2. As i already said we use nimble
3. What part of bt? The host our task or the controller?

**Comment by Platin21** [comment]
And no we have no option to switch to blue droid as that would nog fit in memory anymore.

**Comment by skyESP** [comment]
@Platin21
OK,Thanks for your information. We are trying to reproduce the issue.

Fore the third question. You can do as following.
We have menuconfig to chose running on core0 or core1.
run idf.py menuconfig
For wifi:
(Top) → Component config → Wi-Fi → WiFi Task Core ID
For bt
(Top) → Component config → Bluetooth → Bluetooth controller(ESP32 Dual Mode Bluetooth) → The cpu core which bluetooth controller run

Another two questions
1.What's the priority of the task  for retrieving CAN Messages ? Does it higher than controller task? 
2.From your describe you just use polling mode not using interrupt for handling the task, Am I right?

**Comment by Platin21** [comment]
@skyESP 
Let me check if that changes anything and i get back to you. 

1. 20 from max 25. The BLE one is also slightly evalated i think it’s 15 there.
2. It’s done via Interrupt

**Comment by skyESP** [comment]
OK, You set the  priority of the controller task lower than the task to handle the can message.
Here please do two tests.
1. Set the controller task higher than the  task to handle the can message.If still have problem please do the step 2.
2.Reduce the load or stop the  the  task to handle the can message and check if the issue is still here.
I suspect that the controller task is blocked by other task or the interrupt.

**Comment by Platin21** [comment]
@skyESP where can one change the controller priority? Seems to be not a option inside Component config -> Bluetooth -> Bluetooth Controller. Or what is the default level there?

2. Isn’t really a fix but for testing sure i can do that.

**Comment by skyESP** [comment]
1.You can change the priority in the code where you create the task.
2.From our experience the controller should not be blocked by other tasks or interrupt for long time. Through I do not know how long and how much work you did in your task,so I advance you to do this to  test and verify my suspection. If it works,this can be a fix.

**Comment by Platin21** [comment]
@skyESP you mean the task that handles the BLE so neither the controller nor the host?

**Comment by skyESP** [comment]
@Platin21
hi，
1.You had better keep the priority of controller and host task as our demo example.
2.The task you create yourself should lower than the controller and host ble task.

**Comment by Platin21** [comment]
1. I didn’t change anything there what i was not sure about is if our task should have a higher prio than the can one
2. This was the case anyways i think meanwhile i dunno what the default prio of these other two tasks is. 
3. Possible but that needs to be elaborated in a Email. And i suspect it makes more sense that we setup something as i already said and you maybe can guess that the pcb on which the esp sits is custom.

**Comment by Platin21** [comment]
@skyESP so back to the priorities the host task seems to run on 21 and the LL task seems to run on 24 so our can task is acutally already lower in priority. And to the other tasks on the system all of them run with priority 0 so should not be possbile for it to interfere at all then.

**Comment by skyESP** [comment]
hi,@Platin21
Sorry for replay late. Yes Priority 0 of other tasks should not interfere them.  
For the example  ble_ibeacon demo in our IDF sdk.
The controller task is 23,hci host task  is 22, BTU task is 20 ,BTC task is 19.

**Comment by Platin21** [comment]
Yeah that makes it clearer now. 
We will setup something tomorrow and i will leave my response here. But yeah it sounds very promising. We are now switching the priorities to be under a value of 18. Would be a huge step for us forward if that works out.

**Comment by Platin21** [comment]
After switching the "priority and increasing" the buffer the bug still happens. 
@skyESP Can we arrange a meeting in some ways to get maybe a clearer outcome?

**Comment by Platin21** [comment]
I can share also the crash log again but it is basically the samething as it was.

**Comment by skyESP** [comment]
＠Platin21 
hi,If you stop the task to handle the CAN message. Did this issue still happen?

**Comment by Platin21** [comment]
@skyESP if i stop this task i can’t use CAN so even if it doesn’t happen then that’s not a solution.

**Comment by skyESP** [comment]
Ok,Then what did you do in the interrupt of the CAN? Does this part cost lot of time? Could you share the related code?

**Comment by Platin21** [comment]
Its basically a loop that does rx and tx of can messages and stores them in a static array. 
Looks something like that:

```c++
static void
can_isr( void *arg )
{
  uint8_t events;
  events = can_base->interrupt;

  if( /* check for b0 event */ ) {
    can_rx_isr();
  }
  if(/* check for b1 event */) {
    can_tx_isr();
  }
  /* Check if overrun or bus error occured */
  if( (events & B7) ) {
    can_error_flag = 1u;
    can_error_code_capture = can_base->err_code;
  }
}
```
In side can_rx_isr() is just a function that stores the frame we got and validates it. 
In the tx side we write to the transmit fifo and that seems to be it. (Just a byte copy)

I have to say though without the load of the uds the controller seems to just work fine.

**Comment by Platin21** [comment]
What we are now trying is to make a task that simulates a load and see if that makes any diffrence.

**Comment by skyESP** [comment]
hi,@Platin21 
I think you can post a message from the can_isr to other tasks to  handle the "can_rx_isr()" and "can_tx_isr()".
It should not do lots of work in the interrupt.

**Comment by Platin21** [comment]
@skyESP any estimates how long would be to long to todo something in the interrupt?

**Comment by skyESP** [comment]
The working cases of the controller is complicated(Scan state,adv state,connecting state while receive and transmit data in different frequency), so it is hard to have such data.

**Comment by Platin21** [comment]
@skyESP just a thing on that side the int in our case isn't happening on the same core is a problem with the int then still possible?

**Comment by skyESP** [comment]
Do you mean you can handle the int in another core? I think you can have a try. The ble controller and host is default running on the core 0. You can make the core 1 handle the int. But I am not sure this way can help.

**Comment by ChrisEAlfred** [comment]
We have found that this error is provoked if you regularly use SPIFFS. Maybe during erase cycles.
Unfortunately, without the source code this is almost impossible to catch.

In an earlier post you stated:
"The crash was because after the controller ran out of memory, malloc failure happens and assert is triggered. The issue is usually caused by the slow processing of the application layer, which caused the controller memory could not be released in time"

What is the amount of memory required?
What is the longest "processing of the application layer" that is allowed?

**Comment by Platin21** [comment]
@ChrisEAlfred we are not useing spiffs we are using fat and nvs. In our case basically all memory that the normal esp32 has i think 1k bytes is still left.

Yeah that processing was also interesting for me but it seems like that’s nit easy to say.

**Comment by ChrisEAlfred** [comment]
SPIFFS and NV storage use the same memory.

So I guess the question is ... does using the ESP32 NV in any way provoke this problem?
Our tests appear to indicate that this is the case.

**Comment by Platin21** [comment]
@ChrisEAlfred thing is we didn’t have this controller error when i just run it without having the int from the can driver and after 8 hours still nothing. Teh average time for that to happen was 6 to 4 hours. If it’s the NVS then i’m not sure why this didn’t happen in this case.

**Comment by Platin21** [comment]
@skyESP it seems it's related to our CAN Side. Could it be that somehow something gets overlaped in the sheduler that makes this problem happening?

**Comment by Platin21** [comment]
@ChrisEAlfred we disabled this setting and still get the error so from what i can see there isn't a diffrence.

**Comment by Platin21** [comment]
So for the CAN Side we can't move that to use Polling as that makes the timings not to comply. 
@skyESP any other ideas?

**Comment by Platin21** [comment]
@bshmid i don't think this would be a solution in our case or did you als have troubles when you used the can twai isr source.

**Comment by skyESP** [comment]
@Platin21,I think you can post the work that cost lots of time to one task,the priority of the task should lower than the BTC task.
If you must read or write flash,you should not do it in interrupt.

**Comment by bshmid** [comment]
> @bshmid i don't think this would be a solution in our case or did you als have troubles when you used the can twai isr source. It's all about timings..

I wasn't using the can twai isr source, but I got the error message discussed in this post while running some simple stuff in a while loop.

**Comment by Platin21** [comment]
> > @bshmid i don't think this would be a solution in our case or did you als have troubles when you used the can twai isr source. It's all about timings..
> 
> I wasn't using the can twai isr source, but I got the error message discussed in this post while running some simple stuff in a while loop.

Ok that sounds intressting how "simple" was it in your case? So simple that you'd be able todo a repro so i would also be able to verify if we have this on our board??

**Comment by bshmid** [comment]
Here is it (before fixing):
`  while (1){
  pService->start();
  BLEDevice::startAdvertising();
  Serial.println("Characteristic defined!");
  Serial.print(pServer->getConnectedCount());
  Serial.println(" devices connected");
  
  if (pServer->getConnectedCount()){   //delay(2);
  newString = pCharacteristic->getValue().c_str();
  myString = newString;
  Serial.println(myString);
  val0 = myString.substring(0, 4).toInt();
  val1 = myString.substring(4, 8).toInt();
  val2 = myString.substring(8, 12).toInt();
  val3 = myString.substring(12, 16).toInt();
  dac.analogWrite(dac.A,val0);
  dac.analogWrite(dac.B,val1);
  dac.analogWrite(dac.C,val2);
  dac.analogWrite(dac.D,val3);
  Serial.print(pServer->getConnectedCount());
  Serial.println(" devices connected");
}`

**Comment by Platin21** [comment]
That's it? And how long does it take to fail on your side?

**Comment by bshmid** [comment]
A few seconds.

**Comment by Platin21** [comment]
@bshmid Are these arduino abstractions or did you do them yourselfes? Because I'd like to verify that on our board too.

**Comment by bshmid** [comment]
> @bshmid Are these arduino abstractions or did you do them yourselfes? Because I'd like to verify that on our board too.

Sorry, I don't understand your question.

**Comment by Platin21** [comment]
@bshmid can you provide the full code as in including the setup code?

**Comment by Platin21** [comment]
Ahh got it done myself.

**Comment by Platin21** [comment]
Anything happening on that side? I had to fix a other bug before being able to look at this again. 
@skyESP did you try out the example from @bshmid above?

**Comment by bill1davis** [comment]
Following. idf 4.3.1. ESP32 Chip rev 3.  Saw one of these last night, both ble and wifi running on core 0 connected.  Fairly CPU intensive calculations running on low priority thread on core 1 completed just prior to crash.  SPI bus probably busy switching screens.

ASSERT_PARAM(-218959118 0), in arch_main.c at line 343
I (39889008) ui: onDismissed: 428 : CalibratingScreen
Guru Meditation Error: Core  0 panic'ed (Interrupt wdt timeout on CPU0). 

Core  0 register dump:
PC      : 0x400852a7  PS      : 0x00060c34  A0      : 0x80181940  A1      : 0x3ffbe970  
0x400852a7: r_assert_param at ??:?

A2      : 0x00000001  A3      : 0x00000000  A4      : 0x60008048  A5      : 0x00000000  
A6      : 0x00000004  A7      : 0x3ffbdbe4  A8      : 0x800852a4  A9      : 0x3ffbe950  
A10     : 0x00000000  A11     : 0x00000037  A12     : 0x00000014  A13     : 0xffffffff  
A14     : 0x00000000  A15     : 0xfffffffc  SAR     : 0x00000004  EXCCAUSE: 0x00000005  
EXCVADDR: 0x00000000  LBEG    : 0x400851dc  LEND    : 0x400851e3  LCOUNT  : 0x00000000  
0x400851dc: r_assert_param at ??:?

0x400851e3: r_assert_param at ??:?

Core  0 was running in ISR context:
EPC1    : 0x400d44d3  EPC2    : 0x40234e0a  EPC3    : 0x00000000  EPC4    : 0x400852a7
0x400d44d3: uart_hal_write_txfifo at /home/dev/esp/esp-idf/components/hal/uart_hal_iram.c:35

0x40234e0a: esp_pm_impl_waiti at /home/dev/esp/esp-idf/components/esp_pm/pm_impl.c:829

0x400852a7: r_assert_param at ??:?


Backtrace:0x400852a4:0x3ffbe970 0x4018193d:0x3ffbe990 0x40019fb5:0x3ffbe9b0 0x40089ada:0x3ffbe9e0 0x400883b9:0x3ffbea20 0x400880d1:0x3ffbea40 0x4008a86e:0x3ffbea60 0x4008baeb:0x3ffbea80 0x400847e5:0x3ffbeaa0 0x401f7a11:0x3ffdb5f0 0x401f7a76:0x3ffdb630 0x401b199e:0x3ffdb650 0x402352cd:0x3ffdb670 0x401c3ebf:0x3ffdb690 0x401c38c9:0x3ffdb6b0 0x401be8f5:0x3ffdb6d0 0x401beb86:0x3ffdb700 0x401bf9dd:0x3ffdb730 0x401bfa19:0x3ffdb780 0x401bfa46:0x3ffdb7b0 0x401bbac5:0x3ffdb7e0 0x401ba8b5:0x3ffdb820 0x401bf73a:0x3ffdb890 0x401c3806:0x3ffdb8c0 0x401b436d:0x3ffdb8e0
0x400852a4: r_assert_param at ??:?

0x4018193d: r_platform_reset at ??:?

0x40089ada: r_lld_pdu_rx_handler at ??:?

0x400883b9: r_lld_evt_end at ??:?

0x400880d1: r_lld_evt_end_isr at ??:?

0x4008a86e: r_rwble_isr at ??:?

0x4008baeb: r_rwbtdm_isr_wrapper at intc.c:?

0x400847e5: _xt_lowint1 at /home/dev/esp/esp-idf/components/freertos/port/xtensa/xtensa_vectors.S:1105

0x401f7a11: ieee80211_search_node at ??:?

0x401f7a76: esp_wifi_internal_tx_by_ref at ??:?

0x401b199e: wifi_transmit_wrap at /home/dev/esp/esp-idf/components/esp_wifi/src/wifi_netif.c:70

0x402352cd: esp_netif_transmit_wrap at /home/dev/esp/esp-idf/components/esp_netif/lwip/esp_netif_lwip.c:880

0x401c3ebf: low_level_output at /home/dev/esp/esp-idf/components/lwip/port/esp32/netif/wlanif.c:137

0x401c38c9: ethernet_output at /home/dev/esp/esp-idf/components/lwip/lwip/src/netif/ethernet.c:312 (discriminator 2)

0x401be8f5: etharp_output_to_arp_index at /home/dev/esp/esp-idf/components/lwip/lwip/src/core/ipv4/etharp.c:785

0x401beb86: etharp_output at /home/dev/esp/esp-idf/components/lwip/lwip/src/core/ipv4/etharp.c:901

0x401bf9dd: ip4_output_if_opt_src at /home/dev/esp/esp-idf/components/lwip/lwip/src/core/ipv4/ip4.c:1089

0x401bfa19: ip4_output_if_opt at /home/dev/esp/esp-idf/components/lwip/lwip/src/core/ipv4/ip4.c:890

0x401bfa46: ip4_output_if at /home/dev/esp/esp-idf/components/lwip/lwip/src/core/ipv4/ip4.c:867

0x401bbac5: tcp_output_segment at /home/dev/esp/esp-idf/components/lwip/lwip/src/core/tcp_out.c:1607
 (inlined by) tcp_output at /home/dev/esp/esp-idf/components/lwip/lwip/src/core/tcp_out.c:1360

0x401ba8b5: tcp_input at /home/dev/esp/esp-idf/components/lwip/lwip/src/core/tcp_in.c:557

0x401bf73a: ip4_input at /home/dev/esp/esp-idf/components/lwip/lwip/src/core/ipv4/ip4.c:784

0x401c3806: ethernet_input at /home/dev/esp/esp-idf/components/lwip/lwip/src/netif/ethernet.c:186

0x401b436d: tcpip_thread_handle_msg at /home/dev/esp/esp-idf/components/lwip/lwip/src/api/tcpip.c:180
 (inlined by) tcpip_thread at /home/dev/esp/esp-idf/components/lwip/lwip/src/api/tcpip.c:154

Core  1 register dump:
PC      : 0x40234e0a  PS      : 0x00060b34  A0      : 0x800db280  A1      : 0x3ffca7a0  
0x40234e0a: esp_pm_impl_waiti at /home/dev/esp/esp-idf/components/esp_pm/pm_impl.c:829

A2      : 0x00000000  A3      : 0x400918d0  A4      : 0x00000000  A5      : 0x000000a5  
0x400918d0: _cleanup_r at /builds/idf/crosstool-NG/.build/xtensa-esp32-elf/src/newlib/newlib/libc/stdio/findfp.c:205

A6      : 0x00000014  A7      : 0x00000001  A8      : 0x801c9f22  A9      : 0x3ffca770  
A10     : 0x00000000  A11     : 0x3ffbf0b8  A12     : 0x3ffbf0b8  A13     : 0x00000000  
A14     : 0x00060f20  A15     : 0x3f874ed0  SAR     : 0x00000007  EXCCAUSE: 0x00000005  
EXCVADDR: 0x00000000  LBEG    : 0x00000000  LEND    : 0x00000000  LCOUNT  : 0x00000000  

Backtrace:0x40234e07:0x3ffca7a0 0x400db27d:0x3ffca7c0 0x40098e54:0x3ffca7e0
0x40234e07: cpu_ll_waiti at /home/dev/esp/esp-idf/components/hal/esp32/include/hal/cpu_ll.h:183
 (inlined by) esp_pm_impl_waiti at /home/dev/esp/esp-idf/components/esp_pm/pm_impl.c:827

0x400db27d: esp_vApplicationIdleHook at /home/dev/esp/esp-idf/components/esp_common/src/freertos_hooks.c:63

0x40098e54: prvIdleTask at /home/dev/esp/esp-idf/components/freertos/tasks.c:3839

**Comment by skyESP** [comment]
hi@bill1davis
Please check if there is high level task or interrupt which cost lots of time running. The crash reason is as before."The crash was because after the controller ran out of memory, malloc failure happens and assert is triggered. The issue is usually caused by the slow processing of the high level application task or interrupt, which caused the controller memory could not be released in time."

**Comment by bill1davis** [comment]
hi@skyESP
We have a high priority task running on core 1 that runs for a short time at 10hz.  We have another low priority task on core 1 that does floating point computation that can run for up to 200ms without explicitly giving up its timeslot.  We have a couple of very simple interrupt handlers that just queue events and return.  BLE and Wifi are both configured to run on core 0.

My thinking is that having our demanding threads running on core 1 should minimize interference with the radios.  Do I need to adjust the priority of my high priority task on core 1 below some level?  Thanks!

**Comment by skyESP** [comment]
I think you had better adjust the priority of high priority task on core 1 lower than the BTU and BTC task.

**Comment by ChrisEAlfred** [comment]
In my response https://github.com/espressif/esp-idf/issues/7805#issuecomment-974900198 we reported that using SPIFFS provokes the issue.

We are redirecting the ESP32 logs to SPIFFS using esp_log_set_vprintf. Is it possible this can cause the issue?

**Comment by Platin21** [comment]
As i said ealier we are using FATFS and NVS. And both of these don't use the SPIFFS. As NVS ist just a key value store unformated and FATFS is FAT. So if that is because of this i would suspect that it happens because of some common layer between these systems. But maybe in our case it's just the memory usage that makes problems.

**Comment by guirespi** [comment]
Hi, I'm facing the same issue.

Is there a definitive way to solve this problem?

I get this error when I'm finishing writing in the flash using NVS multiple times.  I got the following logs:

E (165428) task_wdt: Task watchdog got triggered. The following tasks did not reset the watchdog in time:
E (165428) task_wdt:  - IDLE0 (CPU 0)
E (165428) task_wdt: Tasks currently running:
E (165428) task_wdt: CPU 0: btController
E (165428) task_wdt: CPU 1: IDLE1
E (165428) task_wdt: Print CPU 0 (current core) backtrace


Backtrace:0x400D45AC:0x3FFBE4F0 0x40082561:0x3FFBE510 0x400F49A4:0x3FFB4E50 0x400F4F6D:0x3FFB4E70

E (165428) task_wdt: Print CPU 1 backtrace


Backtrace:0x40081966:0x3FFBEAF0 0x40082561:0x3FFBEB10 0x4000BFED:0x3FFBC420 0x4008F911:0x3FFBC430 0x400D47BC:0x3FFBC450 0x400D47C7:0x3FFBC480 0x400D4862:0x3FFBC4A0 0x4008FEA8:0x3FFBC4C0

E (170428) task_wdt: Task watchdog got triggered. The following tasks did not reset the watchdog in time:
E (170428) task_wdt:  - IDLE0 (CPU 0)
E (170428) task_wdt: Tasks currently running:
E (170428) task_wdt: CPU 0: btController
E (170428) task_wdt: CPU 1: IDLE1
E (170428) task_wdt: Print CPU 0 (current core) backtrace


Backtrace:0x400D45AC:0x3FFBE4F0 0x40082561:0x3FFBE510 0x400F49A4:0x3FFB4E50 0x400F4F6D:0x3FFB4E70

E (170428) task_wdt: Print CPU 1 backtrace


Backtrace:0x40081966:0x3FFBEAF0 0x40082561:0x3FFBEB10 0x4000BFED:0x3FFBC420 0x4008F911:0x3FFBC430 0x400D47BC:0x3FFBC450 0x400D47C7:0x3FFBC480 0x400D4862:0x3FFBC4A0 0x4008FEA8:0x3FFBC4C0

E (175428) task_wdt: Task watchdog got triggered. The following tasks did not reset the watchdog in time:
E (175428) task_wdt:  - IDLE0 (CPU 0)
E (175428) task_wdt: Tasks currently running:
E (175428) task_wdt: CPU 0: btController
E (175428) task_wdt: CPU 1: IDLE1
E (175428) task_wdt: Print CPU 0 (current core) backtrace


Backtrace:0x400D45AC:0x3FFBE4F0 0x40082561:0x3FFBE510 0x400F49A4:0x3FFB4E50 0x400F4F6D:0x3FFB4E70

E (175428) task_wdt: Print CPU 1 backtrace


Backtrace:0x40081966:0x3FFBEAF0 0x40082561:0x3FFBEB10 0x4000BFED:0x3FFBC420 0x4008F911:0x3FFBC430 0x400D47BC:0x3FFBC450 0x400D47C7:0x3FFBC480 0x400D4862:0x3FFBC4A0 0x4008FEA8:0x3FFBC4C0

E (180428) task_wdt: Task watchdog got triggered. The following tasks did not reset the watchdog in time:
E (180428) task_wdt:  - IDLE0 (CPU 0)
E (180428) task_wdt: Tasks currently running:
E (180428) task_wdt: CPU 0: btController
E (180428) task_wdt: CPU 1: IDLE1
E (180428) task_wdt: Print CPU 0 (current core) backtrace


Backtrace:0x400D45AC:0x3FFBE4F0 0x40082561:0x3FFBE510 0x400F49A4:0x3FFB4E50 0x400F4F6D:0x3FFB4E70

E (180428) task_wdt: Print CPU 1 backtrace


Backtrace:0x40081966:0x3FFBEAF0 0x40082561:0x3FFBEB10 0x4000BFED:0x3FFBC420 0x4008F911:0x3FFBC430 0x400D47BC:0x3FFBC450 0x400D47C7:0x3FFBC480 0x400D4862:0x3FFBC4A0 0x4008FEA8:0x3FFBC4C0

ASSERT_PARAM(-218959118 0), in arch_main.c at line 343
Guru Meditation Error: Core  0 panic'ed (Interrupt wdt timeout on CPU0).

I must say in my program I'm using ISR to replicate the example of espressif called "timer_group_example_main". I use this example to implement a timer and a tick counter, but, when I use this, this error appears and crash the program and doesn't restart until I press the EN button in the ESP-WROVER-KIT. I must say that when I disable BLE_MODEM_SLEEP the error doesn't appear immediately but randomly after several hours.

**Comment by guirespi** [comment]
@Platin21 Hi, you never could find the solution to this bug?

**Comment by Platin21** [comment]
Yes we never got a reply in the sense of this is how one can fix it. Or these are conditions where this happens. We also ended long term testing some months ago as we can’t do anything without getting help from espressif. But we will have todo these tests again anyways as the device needs for certification a minimum runtime of 16 hours. We are using a ublox variant of the esp32 to make that also clear.

**Comment by Platin21** [comment]
@guirespi are you able to make a minimal repro? Sad thing is also that there are people just writing in this issue there own stuff which is completely unrelated so the conversation driftet to there off topic problems.

**Comment by guirespi** [comment]
@Platin21 I will try it this week from the blephr and timer_group_example example and do some task that writes and erase the flash periodically. I think those are the conditions to trigger the wdt timeout.

**Comment by AxelLin** [comment]
> I think you had better adjust the priority of high priority task on core 1 lower than the BTU and BTC task.

@skyESP 
But you cannot complete control the task priority because the assigned priorities of tasks shifts after creation. see https://github.com/espressif/esp-idf/issues/7580

**Comment by Platin21** [comment]
So after sometime not having any time to test this we found that the bug still happens. We where thinking that it doesn't happen anymore after we changed some stuff on the nimble side. This turned out to be wrong. We still have the same problem there. In this case we where using the idf version 4.3.2 there is no option just by the flash size too switch to a newer version. So the question is did anything change that could change the beavhoiur of the libbtm in newer versions?

Basically the same as above:

ASSERT_PARAM(-218959118 0), in arch_main.c at line 358
Guru Meditation Error: Core  0 panic'ed (IllegalInstruction). Exception was unhandled.
Memory dump at 0x40083100: 0000f01d 00004136 f01d0000
Core  0 register dump:
PC      : 0x40083107  PS      : 0x00060133  A0      : 0x800834e5  A1      : 0x3ffb5ce0
A2      : 0x00000000  A3      : 0xf2f2f2f2  A4      : 0x00000000  A5      : 0x3f414e66
A6      : 0x00000166  A7      : 0xfffffffc  A8      : 0x8000814b  A9      : 0x3ffb5c50
A10     : 0x00000000  A11     : 0x3ffb5c73  A12     : 0x3ffb5c1f  A13     : 0x00000038
A14     : 0x00000000  A15     : 0x3ffb5c24  SAR     : 0x00000004  EXCCAUSE: 0x00000000
EXCVADDR: 0x00000000  LBEG    : 0x400833ed  LEND    : 0x400833f5  LCOUNT  : 0x00000000
Backtrace:0x40083104:0x3ffb5ce0 0x400834e2:0x3ffb5d00 0x4010cb3d:0x3ffb5d20 0x40019fb5:0x3ffb5d40 0x4001a1f2:0x3ffb5d70 0x4011aebf:0x3ffb5d90 0x4011c629:0x3ffb5e00 0x40119737:0x3ffb5e50 0x401169ba:0x3ffb5ea0 0x40019d11:0x3ffb5ee0 0x40055b4d:0x3ffb5f00 0x4010cf47:0x3ffb5f20 0x4010d5cc:0x3ffb5f40
ELF file SHA256: 059dc4e8a0e2b470
Rebooting...

Decoded Lines:
r_assert
??:?
r_assert_param
??:?
r_platform_reset
??:?
??
??:0
??
??:0
r_llm_le_adv_report_ind
??:?
r_llm_pdu_defer
??:?
r_lld_pdu_check
??:?
r_lld_evt_deffered_elt_handler
??:?
??
??:0
??
??:0
r_rw_schedule
??:?
btdm_controller_task
??:?


The CPU workload is in our setup only using 30% with two cores active + 240mhz. So from a load side i can't see a reason why. 
Memory is one point that we have a relativly high usage of 70% of the available heap. with a freespace of 41,4 KB. 

Basically meaning that we neither have no memory nor is the cpu doing a lot of work. Which makes the load as a cause not really plausible. So any other ideas?
@skyESP @TianaESP

**Comment by Platin21** [comment]
We are now also testing with 4.4.1 to see if it still happens.

**Comment by Platin21** [comment]
After testing with the above mentioned Version we still have the same Problem so the current IDF doesn't fix it:

ASSERT_PARAM(-218959118 0), in arch_main.c at line 358
Guru Meditation Error: Core  1 panic'ed (IllegalInstruction). Exception was unhandled.
Memory dump at 0x40090094: 000000f0 00004136 f01d0000
Core  1 register dump:
PC      : 0x4009009b  PS      : 0x00060c34  A0      : 0x80083f4d  A1      : 0x3ffb5ed0
A2      : 0x00000000  A3      : 0xf2f2f2f2  A4      : 0x00000000  A5      : 0x3f412c66
A6      : 0x00000166  A7      : 0xfffffffc  A8      : 0x8000814b  A9      : 0x3ffb5e40
A10     : 0x00000000  A11     : 0x3ffb5e63  A12     : 0x3ffb5e0f  A13     : 0x00000038
A14     : 0x00000000  A15     : 0x3ffb5e14  SAR     : 0x00000004  EXCCAUSE: 0x00000000
EXCVADDR: 0x00000000  LBEG    : 0x40083e55  LEND    : 0x40083e5d  LCOUNT  : 0x00000000
Backtrace:0x40090098:0x3ffb5ed00x40083f4a:0x3ffb5ef0 0x4010b189:0x3ffb5f10 0x40019fb5:0x3ffb5f30 0x4001a1f2:0x3ffb5f60 0x4011987b:0x3ffb5f80 0x4011afe5:0x3ffb5ff0 0x401180f3:0x3ffb6040 0x40115372:0x3ffb6090 0x40019d11:0x3ffb60d0 0x40055b4d:0x3ffb60f0 0x4010b58b:0x3ffb6110 0x4010bc10:0x3ffb6130
ELF file SHA256: 5e1605f6b8f2252a
Rebooting...

**Comment by Platin21** [comment]
So any other leads? NVS didn't do the trick, we also don't have a lot of load and there is plenty of memory left as in 42kb memory. 

And here is the decoded output:

r_assert
C:/Espressif_4_4_1/frameworks/esp-idf-v4.4.1/components/bt/controller/esp32/bt.c:2018
r_platform_reset
??:?
??
??:0
??
??:0
r_llm_le_adv_report_ind
??:?
r_llm_pdu_defer
??:?
r_lld_pdu_check
??:?
r_lld_evt_deffered_elt_handler
??:?
??
??:0
??
??:0
r_rw_schedule
??:?
btdm_controller_task
??:?

**Comment by AxelLin** [comment]
> @Alvin1Zhang This issue has been reported quite long time ago, and at least 3 people reported that the fix does not really fix the issue. [#6517 (comment)](https://github.com/espressif/esp-idf/issues/6517#issuecomment-891186177)

@skyESP @TianaESP 
Any update/fix for this issue?

**Comment by W-M-D** [comment]
EEEK I just had this happen. The worst part is it doesn't call abort so my motors were still running and it caused a crash.

**Comment by esp-cjh** [comment]
Could you please provide the completed log? I need to known the controller compile version.

**Comment by esp-cjh** [comment]
We have fixed this bug(commit ed7d2d3f0db5fcb626a3ebb799b9bcce66f674e4 for branch release/v4.4). Please update the code.

**Comment by guirespi** [comment]
Hi, after a long time of testing I can say that this error happens when a long time operation (like saving a blog larger than 2k bytes in the nvs) and BLE is receiving or writing data. It still happens in ESP IDF v4.4.2, @esp-cjh. I saw that the bugfix was made in the esp32c3 family... what about the others?

**Comment by wilsonnggp** [comment]
Anyone have any finite fixes for this? 

![Screen Shot 2023-03-13 at 11 16 54 AM](https://user-images.githubusercontent.com/99208898/224794261-2e2eebfa-b66c-4220-870b-2db76cb1bd94.png)

I'm using WiFi, BLE, TWAI, and SPIFFS and some print logs my application.

OTA works 2-3 times consistently and then the 4th attempt it usually just crashes. I can also comment to @guirespi  that I'm also using storage that is 1MB which may be a factor on the memory corruption. However I'm always un mounting, registering the drive before accessing as well as a clean format once I load the OTA application upon it.

**Comment by bill1davis** [comment]
We havent seen it since auditing our high priority tasks and making sure they are lower than btu and btc as @skyESP  suggested.

The task in question was doing almost nothing, but it made a difference that it was preempting the bt tasks, I guess.

**Comment by esp-cjh** [comment]
@guirespi 
We have fixed this bug (commit 8769f62c3b687a4dedf25c748fe040b47ea6b2d4 for branch release/v4.3). Please cherry-pick the commit and see if problems also arise.

**Comment by ianzo** [comment]
@esp-cjh We tested the idf 4.4.4 where the fix is present but the problem still happen while writing on FLASH. The target device is the esp32. 

See the Backtrace below.

ASSERT_PARAM(-218959118 0), in arch_main.c at line 358
Guru Meditation Error: Core  1 panic'ed (IllegalInstruction). Exception was unhandled.
Memory dump at 0x4008fda4: 000000f0 00004136 f01d0000
0x4008fda4: coex_bb_reset_unlock_wrapper at /opt/esp/idf/components/bt/controller/esp32/bt.c:1302

Core  1 register dump:
PC      : 0x4008fdab  PS      : 0x00060734  A0      : 0x80083e11  A1      : 0x3ffc1320
0x4008fdab: r_assert at /opt/esp/idf/components/bt/controller/esp32/bt.c:2019

A2      : 0x00000000  A3      : 0xf2f2f2f2  A4      : 0x00000000  A5      : 0x3f433b26
A6      : 0x00000166  A7      : 0xfffffffc  A8      : 0x8000814b  A9      : 0x3ffc1290
A10     : 0x00000000  A11     : 0x3ffc12b3  A12     : 0x3ffc125f  A13     : 0x00000038
A14     : 0x00000000  A15     : 0x3ffc1264  SAR     : 0x00000004  EXCCAUSE: 0x00000000
EXCVADDR: 0x00000000  LBEG    : 0x40083d19  LEND    : 0x40083d21  LCOUNT  : 0x00000000
0x40083d19: r_assert_param at ??:?

0x40083d21: r_assert_param at ??:?



Backtrace:0x4008fda8:0x3ffc13200x40083e0e:0x3ffc1340 0x40143d39:0x3ffc1360 0x40019fb5:0x3ffc1380 0x4008862a:0x3ffc13b0 0x40086f25:0x3ffc13f0 0x40086c3e:0x3ffc1410 0x4008940a:0x3ffc1430 0x4008a673:0x3ffc1450 0x400832d5:0x3ffc1470 0x40083242:0x3ffc1490 0x40083226:0x00000000  |<-CORRUPTED
0x4008fda8: r_assert at /opt/esp/idf/components/bt/controller/esp32/bt.c:2018

0x40083e0e: r_assert_param at ??:?

0x40143d39: r_platform_reset at ??:?

0x4008862a: r_lld_pdu_rx_handler at ??:?

0x40086f25: r_lld_evt_end at ??:?

0x40086c3e: r_lld_evt_end_isr at ??:?

0x4008940a: r_rwble_isr at ??:?

0x4008a673: r_rwbtdm_isr_wrapper at intc.c:?

0x400832d5: hli_c_handler at /opt/esp/idf/components/bt/controller/esp32/hli_api.c:101

0x40083242: _highint4_stack_switch at /opt/esp/idf/components/bt/controller/esp32/hli_vectors.S:185

0x40083226: xt_highint4 at /opt/esp/idf/components/bt/controller/esp32/hli_vectors.S:169

**Comment by esp-cjh** [comment]
@ianzo OK. Could you please provide the procedure or source code to reproduce this issue?

**Comment by guirespi** [comment]
Hi @esp-cjh, long time ago we avoid this bug disabling the option in SDKConfig "Enable Bluetooth modem sleep", but the bug was triggered when I was writting to the flash a chunk of 4 kB using your nvs_flash library and then connecting to BLE, with modem sleep enable, and writting to a  characteristic some value that will be written into the flash too. 

It is curious that this only happen if the option of modem sleep is enable and without it I was able to write the chunk of 4096 B and also write the value wrote through the BLE characteristic. By the way, this happened to me while using ESP IDF 4.4.2.

P.S. after this bug was avoided, our device was able to write to flash chunks of 4096 periodically, communicate with a sensor via UART and write its value into flash each 220 ms, all this coexisting with BLE up to 21 consecutive days without resets. (A blackout cause the restart of the device at the end)

**Comment by esp-cjh** [comment]
@guirespi Thanks. We will reproduce the bug and fixed it ASAP.

**Comment by esp-cjh** [comment]
@guirespi How often do you write flash? How often do you write GATT characteristic value?

**Comment by guirespi** [comment]
Hi @esp-cjh, the writing in the GATT characteristic was user choice. The problem triggers when the user writes and the chunk of 4096 bytes is written too by a task. I was able to make this by setting the local time in the esp32 with a server and then schedule the writing by a specific hour. E.g. every quarter hour the chunk is written and just when this starts, I write to the characteristic.

**Comment by esp-cjh** [comment]
@guirespi I write characteristic value all the time and write flash (4096 bytes) every 2 seconds. But can't reproduce the problem. Is there something wrong with my reproduction? By the way, the task used to write flash has a low priority.

**Comment by guirespi** [comment]
@esp-cjh is the write to the characteristic at the same as the writing of the flash? Just to be clear, in my example I wrote to a gatt characteristic 500 bytes and those bytes are written to the flash. At the same time, the part of 4096 is written too. The two byte's arrays correspond to different key-values pairs and were saved using the nvs api.

About your question, the ble task had the highest priority and the the other task has a value near it. I don't have right now the exact  values but if I remember right it was 22 and 16 respectively. Also, the option "ble modem sleep" was enable when this bug happened.

I hope this helps to reproduce the bug, sorry if it doesn't work or I'm misunderstanding the problem.

**Comment by guirespi** [comment]
@esp-cjh I forgot about something. When I said the 500 bytes of the gatt characteristic are written to the flash, is more like that from the 500 bytes, I separate those in 10 parts, and those 10 parts are written in different key-value pairs in a for loop. And then I notify to the user an 'OK' via the same characteristic.

**Comment by guirespi** [comment]
@esp-cjh the last detail haha Each write of a key-value pair creates a different nvs handle and when the change is commited, closes it. If other key-value has to be written, then I create another nvs handle and so on.

The chunk of 4096 bytes also creates its own nvs handle when try to write to the flash and then closes it when finish.

**Comment by ianzo** [comment]
> @ianzo OK. Could you please provide the procedure or source code to reproduce this issue?

@esp-cjh if you make a task that reads from FLASH through the Bluetooth characteristics meanwhile another task is writing in FLASH, then you can reproduce the bug. Hereafter there is the Backtrace that replicates the problem and the issue is present with idf 5.0.2 as well! 

Backtrace: 0x4016C272:0x3FFBE6F0 0x40082E15:0x3FFBE710 0x4008C551:0x3FFCD9E0 0x4008D003:0x3FFCDA10 0x40168399:0x3FFCDA80 0x40130F71:0x3FFCDAB0 0x4013487F:0x3FFCDAD0 0x401324DE:0x3FFCDB00 0x40133BEB:0x3FFCDB50 0x40131411:0x3FFCDB90 0x40130D68:0x3FFCDBC0 0x400D6723:0x3FFCDC10 0x401FD504:0x3FFCDC30 0x401FD5AD:0x3FFCDC60 0x400E43EA:0x3FFCDC80 0x400E5262:0x3FFCDD60 0x400E58B9:0x3FFCDDB0 0x400ECA71:0x3FFCDE10 0x400EB7E0:0x3FFCDE50 0x400EBD93:0x3FFCDF00 0x400EA9A5:0x3FFCDF30 0x400DCB02:0x3FFCDF80 0x400D9B2C:0x3FFCDFC0
0x4016c272: task_wdt_isr at /opt/esp/idf/components/esp_system/task_wdt.c:183

0x40082e15: _xt_lowint1 at /opt/esp/idf/components/freertos/port/xtensa/xtensa_vectors.S:1114

0x4008c551: spi_flash_disable_interrupts_caches_and_other_cpu at /opt/esp/idf/components/spi_flash/cache_utils.c:165 (discriminator 1)

0x4008d003: spi_flash_guard_start at /opt/esp/idf/components/spi_flash/flash_ops.c:243
(inlined by) spi_flash_read at /opt/esp/idf/components/spi_flash/flash_ops.c:709

0x40168399: esp_partition_read at /opt/esp/idf/components/spi_flash/partition.c:426

0x40130f71: spiffs_api_read at /opt/esp/idf/components/spiffs/spiffs_api.c:36

0x4013487f: spiffs_phys_rd at /opt/esp/idf/components/spiffs/spiffs/src/spiffs_cache.c:161

0x401324de: spiffs_obj_lu_find_entry_visitor at /opt/esp/idf/components/spiffs/spiffs/src/spiffs_nucleus.c:166

0x40133beb: spiffs_object_find_object_index_header_by_name at /opt/esp/idf/components/spiffs/spiffs/src/spiffs_nucleus.c:1711

0x40131411: SPIFFS_open at /opt/esp/idf/components/spiffs/spiffs/src/spiffs_hydrogen.c:230 (discriminator 2)

0x40130d68: vfs_spiffs_open at /opt/esp/idf/components/spiffs/esp_spiffs.c:523

0x400d6723: esp_vfs_open at /opt/esp/idf/components/vfs/vfs.c:399 (discriminator 3)

0x401fd504: _fopen_r at /builds/idf/crosstool-NG/.build/xtensa-esp32-elf/src/newlib/newlib/libc/stdio/fopen.c:129

0x401fd5ad: fopen at /builds/idf/crosstool-NG/.build/xtensa-esp32-elf/src/newlib/newlib/libc/stdio/fopen.c:168

0x400e43ea: config_check_params_def(web_api_arg_s*) at /project/esp/main/webserver.cpp:1350

0x400e5262: web_call_func(httpd_req*, web_api const*, web_api_arg_s*, int, char const**) at /project/esp/main/webserver.cpp:3776

0x400e58b9: all_api_handler(httpd_req*) at /project/esp/main/webserver.cpp:3848

0x400eca71: httpd_uri at /project/esp/main/esp_http_server/src/httpd_uri.c:313

0x400eb7e0: httpd_parse_req at /project/esp/main/esp_http_server/src/httpd_parse.c:645
(inlined by) httpd_req_new at /project/esp/main/esp_http_server/src/httpd_parse.c:715

0x400ebd93: httpd_sess_process at /project/esp/main/esp_http_server/src/httpd_sess.c:354

0x400ea9a5: httpd_server at /project/esp/main/esp_http_server/src/httpd_main.c:319

0x400dcb02: select_loop at /project/esp/main/select_mgmt.c:139

0x400d9b2c: appmain_task at /project/esp/main/dabmgr.c:134

 

Backtrace: 0x4008D6B5:0x3FFBECF0 0x40082E15:0x3FFBED10 0x4000BFED:0x3FFBD260 0x400954D2:0x3FFBD270 0x4016C5DB:0x3FFBD290 0x4016C5E7:0x3FFBD2B0 0x400D5B62:0x3FFBD2D0 0x40092E03:0x3FFBD2F0
0x4008d6b5: esp_crosscore_isr at /opt/esp/idf/components/esp_system/crosscore_int.c:92

0x40082e15: _xt_lowint1 at /opt/esp/idf/components/freertos/port/xtensa/xtensa_vectors.S:1114

0x400954d2: vPortClearInterruptMaskFromISR at /opt/esp/idf/components/freertos/port/xtensa/include/freertos/portmacro.h:571
(inlined by) vPortExitCritical at /opt/esp/idf/components/freertos/port/xtensa/port.c:332

0x4016c5db: esp_task_wdt_reset at /opt/esp/idf/components/esp_system/task_wdt.c:330

0x4016c5e7: idle_hook_cb at /opt/esp/idf/components/esp_system/task_wdt.c:80

0x400d5b62: esp_vApplicationIdleHook at /opt/esp/idf/components/esp_system/freertos_hooks.c:51 (discriminator 1)

0x40092e03: prvIdleTask at /opt/esp/idf/components/freertos/tasks.c:3987

 
ASSERT_PARAM(-218959118 0), in arch_main.c at line 364
Guru Meditation Error: Core  0 panic'ed (IllegalInstruction). Exception was unhandled.
Memory dump at 0x4008fe2c: 000000f0 00004136 f01d0000
0x4008fe2c: coex_bb_reset_unlock_wrapper at /opt/esp/idf/components/bt/controller/esp32/bt.c:1161

Core  0 register dump:
PC      : 0x4008fe33  PS      : 0x00060c34  A0      : 0x80083f45  A1      : 0x3ffc1338
0x4008fe33: r_assert at /opt/esp/idf/components/bt/controller/esp32/bt.c:1868

A2      : 0x00000000  A3      : 0xf2f2f2f2  A4      : 0x00000000  A5      : 0x3f434846
A6      : 0x0000016c  A7      : 0xfffffffc  A8      : 0x8000814b  A9      : 0x3ffc12a8
A10     : 0x00000000  A11     : 0x3ffc12cb  A12     : 0x3ffc1277  A13     : 0x00000034
A14     : 0x00000000  A15     : 0x3ffc127c  SAR     : 0x00000004  EXCCAUSE: 0x00000000
EXCVADDR: 0x00000000  LBEG    : 0x40083e4d  LEND    : 0x40083e55  LCOUNT  : 0x00000000
0x40083e4d: r_assert_param at ??:?

0x40083e55: r_assert_param at ??:?

Backtrace: 0x4008fe30:0x3ffc1338 |<-CORRUPTED
0x4008fe30: r_assert at /opt/esp/idf/components/bt/controller/esp32/bt.c:1867

**Comment by zhp0406** [comment]
@ianzo 
Have you resolved the issue? could you provide the code to help me reproduce the issue.

**Comment by Alvin1Zhang** [comment]
Thanks for reporting, would you please help share if any further updates for the issue? Thanks.

**Comment by zhp0406** [comment]
fixed commit:
master 9a46d06532d27954af138cd40e36c599681bb62f
v4.4 f39df9a9c002066b4d2cbc26f086dd5738d94b25
v4.3 8769f62c3b687a4dedf25c748fe040b47ea6b2d4

**Comment by Alvin1Zhang** [comment]
Thanks for reporting, will close due to inactivity, feel free to reopen if the issue still happens.
